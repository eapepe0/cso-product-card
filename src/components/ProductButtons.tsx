import React ,{ useContext , useCallback } from "react";
import { ProductContext } from "./ProductCard";

import styles from '../styles/styles.module.css';

export interface Props{
    className? : string;
    style? : React.CSSProperties;
}

export const ProductButtons = ({className, style} : Props) =>{

    const { increaseBy , counter , maxCount } = useContext(ProductContext)

   const isMaxReached = useCallback(
     () => !!maxCount && counter === maxCount ,
     [counter , maxCount],
   )

    return (
    <div className={`${styles.buttonsContainer} ${className}`} style={style}>
        <button className={styles.buttonMinus} onClick={() => increaseBy(-1)}>
            -
        </button>
        <div className={styles.countLabel}>{counter}</div>
        <button className={`${styles.buttonAdd} ${isMaxReached () && styles.disabled}`} onClick={() => increaseBy(+1)}>
            +
        </button>
    </div>
   );
}
/**
 * //* linea 4 : importamos los estilos
 * //* linea 6 : creamos una interface para recibir el className , el cual definimos como que puede existir y es una string
 * 
 * //* linea 10 : pasamos el className y le pasamos la interface 
 * 
 * //* linea 15 : le aplicamos los estilos pasados por el className el cual agregamos con los backticks a los styles del propio componente
 * //*            se lo aplicamos al contenedor asi , se aplican a todos los elementos hijos , en caso de querer aplicar solo a un elemento
 * //*            se usa en el css .custom-css button{ color : red } , se define el elemento en este caso button
 */