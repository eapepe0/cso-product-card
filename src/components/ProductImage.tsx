import styles from '../styles/styles.module.css';
import noImage from "../assets/no-image.jpg";
import React ,{ useContext } from "react";
import { ProductContext } from "./ProductCard";

export interface PropsImage{
    img? : string;
    className? : string;
    style? : React.CSSProperties;
}
export const ProductImage = ({ img , className , style } : PropsImage)=>{

    const { product } = useContext(ProductContext)

    let imgToShow :string; 

    if(img){ 
      imgToShow = img;
    } else if (product.img){ 
      imgToShow = product.img;
    }else { 
      imgToShow = noImage;
    }

    return(
        <img className={`${styles.productImg} ${className}`} src={imgToShow} alt="Product" style={style}/>
    )
}


/**
 * //* linea 0  : dividimos la imagen en otro peque√±o componente para poderle dar mas control al usuario
 * 
 * //* linea 6 : creamos una interface donde recibimos una imagen que puede existir o no que es un string y un className que puede existir o no que es una string
 * 
 * //* linea 9  : extraemos el product del context
 * //* linea 10 : le pasamos un img ,y un className con la interface PropsImage
 * //* linea 11 : definimos que la variable es un string
 * 
 * //* linea 13 : si existe una imagen como prop
 * //* linea 15 : si existe el product.img mostramos eso
 * //* linea 17 : si no cae en ninguna de las anteriores
 * 
 *  //* linea 22 : le decimos al src que si existe el img , lo usamos de lo contrario usamos el noImage
 */