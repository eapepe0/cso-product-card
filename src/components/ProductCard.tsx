import styles from '../styles/styles.module.css';
import React ,{createContext} from 'react'
import { useProduct } from "../hooks/useProduct";

import { InitialValues, onChangeArgs, Product, ProductCardHandlers, ProductContextProps } from '../interfaces/interfaces';
import { ProductTitle } from './ProductTitle';
import { ProductImage } from './ProductImage';
import { ProductButtons } from './ProductButtons';

export interface Props {
  product: Product;
  children : (args : ProductCardHandlers)=>JSX.Element;
  className? : string;
  style? : React.CSSProperties;
  onChange?: ( args : onChangeArgs )=> void;
  value? : number;
  initialValues : InitialValues ;
}

export const ProductContext = createContext ({} as ProductContextProps); 
const { Provider } = ProductContext;

export const ProductCard = ({ children,product, className,style , onChange , value , initialValues   }: Props) => {

  const { counter, increaseBy , maxCount , isMaxCountReached , reset } = useProduct({ onChange , product , value , initialValues }); 

  return (
    <Provider value={{counter , increaseBy, product , maxCount }}>
      <div className={`${styles.productCard} ${className}`} style={style}>
        {
          children({
            count : counter,
            isMaxCountReached,
            maxCount : initialValues?.maxCount,
            product,
            increaseBy ,
            reset,
          })
        }
      </div>
    </Provider>
  );
};


ProductCard.Title   = ProductTitle;   //* importamos ProductTitle de su componente
ProductCard.Image   = ProductImage;   //* importamos ProductImage de su componente
ProductCard.Buttons = ProductButtons; //* importamos ProductButtons de su componente


/******************
* //* linea 10 : definimos la interface de ProductCard
* //* linea 11 : recibira un product
* //* linea 12 : recibimos un children , el cual es un Elemento React que puede ser uno solo o varios (un arreglo)
* //* linea 13 : definimos que reciba un className el cual sera opcional y sera una string
* //* liena 14 : definimos un estilo , que puede existir o no , la cual es una React.CSSProperties
* //* linea 15 : definimos una funcion onChange , que puede existir o no , la cual es una funcion que no devuelve nada 
* //*            recibe unos argumentos , onChangeArgs definidos en el interfaces.ts
* 
* //* linea 18 :  creamos un contexto para compartir entre padre e hijo
* //*              con el (as) decimos como se vera el context mediante la interface
* //*              lo exportamos asi lo pueden usar los otros componentes
********************  
* //* linea 21 : vamos a recibir el children , el product y opcionalmente el className , un style y un onChange
*
* //* linea 23 : extraemos del hook useProduct el counter y la funcion increaseBy y le daremos al hook useProduct una funcion onChange
*
* //* linea 26 - 30 : envolvemos todo con el Provider en el value van los valores , los objetos y las funciones que vamos a compartir
* //* linea 27      : concatenamos los estilos propios del componente , y le agregamos unos via className
* //* linea 28      : aca van los otros componentes los cuales vamos a recibir como hijos
 *********************/