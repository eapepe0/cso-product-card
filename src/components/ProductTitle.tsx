import React ,{ useContext } from "react";
import { ProductContext } from "./ProductCard";

import styles from '../styles/styles.module.css';

export interface Props {
    title? : string;
    className? : string;
    style? : React.CSSProperties;
}

export const ProductTitle= ({title , className , style} : Props) =>{

    const { product } = useContext(ProductContext);

      return(
          <span className={`${styles.productDescription} ${className}`} style={style}>{title ? title : product.title}</span>
      )
}

/**
 * //* linea 6 : recibimos un titulo el cual es una string , pero puede no existir y un className que tambien puede o no existir y es una string
 * //* creamos una interface por que le estamos agregando el className 
 * 
 * //* linea 13 : extraemos el product del context
 * 
 *//* linea 16 : si existe el titulo lo mostramos sino el product.title , concatenamos el estilo definido y uno creado por nosotros
 */