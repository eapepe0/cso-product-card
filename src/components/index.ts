import { ProductCard as ProductCardHOC } from "./ProductCard"; 

import { ProductButtons } from "./ProductButtons";
import { ProductImage } from "./ProductImage";
import { ProductTitle } from "./ProductTitle";
import { ProductCardHOCProps } from "../interfaces/interfaces";



export { ProductButtons } from "./ProductButtons";
export { ProductImage } from "./ProductImage";
export { ProductTitle } from "./ProductTitle";

export const ProductCard: ProductCardHOCProps = Object.assign(ProductCardHOC , {
    Title   : ProductTitle , 
    Image   : ProductImage ,
    Buttons : ProductButtons,
})  

export default ProductCard;
/**
 * //* linea 1 : importamos el componente ProductCard y lo renombramos como ProductCardHOC (high order component)
 * 
 * //* linea 14 : ProductCard ahora es un objeto , lo cual tiene que tener un titulo , una imagen y unos botones
 * //*           en JavaScript todo es un objeto excepto los primitivos , por eso le agregaremos al objeto ProductCard nuevas propiedades
 * //*           usaremos la interface creada en el archivo interfaces.ts
 * 
 * //* linea 15 - 16 - 17 : Lo ponemos en mayuscula por que es un componente
 * 
 */