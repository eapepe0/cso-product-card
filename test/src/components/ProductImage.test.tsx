import React from 'react';
import renderer from 'react-test-renderer';

import { ProductCard } from '../../../src/components/ProductCard';
import { ProductImage } from '../../../src/components/ProductImage';

import { product2 } from '../data/products';

describe('ProductImage', () => {
  test('debe de mostar el componente correctamente con el titulo personalizado', () => {
    const wrapper = renderer.create(
      <ProductImage img="../../../src/assets/no-image.jpg" />
    );

    console.log(wrapper.toJSON());
    expect(wrapper.toJSON()).toMatchSnapshot();
  });

  test('debe de mostrar el componente con la imagen del producto', () => {
    const wrapper = renderer.create(
      <ProductCard
        product={product2}
        initialValues={{
          count: 4,
          maxCount: 10,
        }}
      >
        {() => <ProductImage />}
      </ProductCard>
    );

    console.log(wrapper.toJSON());
    expect(wrapper.toJSON()).toMatchSnapshot();
  });
});
