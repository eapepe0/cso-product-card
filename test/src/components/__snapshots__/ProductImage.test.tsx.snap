// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductImage debe de mostar el componente correctamente con el titulo personalizado 1`] = `
<img
  alt="Product"
  className="productImg undefined"
  src="../../../src/assets/no-image.jpg"
/>
`;

exports[`ProductImage debe de mostrar el componente con el nombre del producto 1`] = `
<div
  className="productCard undefined"
>
  <img
    alt="Product"
    className="productImg undefined"
    src="./coffee-mug2.png"
  />
</div>
`;

exports[`ProductImage debe de mostrar el componente con la imagen del producto 1`] = `
<div
  className="productCard undefined"
>
  <img
    alt="Product"
    className="productImg undefined"
    src="./coffee-mug2.png"
  />
</div>
`;
